src/util/live_performance.js

```diff
@@ -0,0 +1,137 @@
+// @flow
+
+import window from './window.js';
+import {version as sdkVersion} from '../../package.json';
+import type Projection from '../geo/projection/projection.js';
+import {
+    isMapboxHTTPStyleURL,
+    isMapboxHTTPTileJSONURL,
+    isMapboxHTTPSpriteURL,
+    isMapboxHTTPFontsURL
+} from './mapbox.js';
+
+type LivePerformanceMetrics = {
+    counters: Array<Object>,
+    metadata: Array<Object>,
+    attributes: Array<Object>
+};
+
+export type LivePerformanceData = {
+    interactionRange: [number, number],
+    width: number,
+    height: number,
+    terrainEnabled: boolean,
+    fogEnabled: boolean,
+    projection: Projection,
+    renderer: ?string,
+    vendor: ?string
+};
+
+function categorize(arr, fn) {
+    const obj = {};
+    if (arr) {
+        for (const item of arr) {
+            const category = fn(item);
+            if (obj[category] === undefined) {
+                obj[category] = [];
+            }
+            obj[category].push(item);
+        }
+    }
+    return obj;
+}
+
+function getTransferRangePerResourceType(resourceTimers) {
+    const obj = {};
+    if (resourceTimers) {
+        for (const category in resourceTimers) {
+            if (category !== 'other') {
+                for (const timer of resourceTimers[category]) {
+                    const min = `${category}TransferStart`;
+                    const max = `${category}TransferEnd`;
+
+                    // Resource -TransferStart and -TransferEnd represent the wall time
+                    // between the start of a request to when the data is available
+                    obj[min] = Math.min(obj[min] || +Infinity, timer.startTime);
+                    obj[max] = Math.max(obj[max] || -Infinity, timer.responseEnd);
+                }
+            }
+        }
+    }
+    return obj;
+}
+
+function getResourceCategory(entry: PerformanceResourceTiming): string {
+    const url = entry.name.split('?')[0];
+
+    // Code may be hosted on various endpoints: CDN, self-hosted,
+    // from unpkg... so this check doesn't include mapbox HTTP URL
+    if (url.includes('mapbox-gl.js')) return 'javascript';
+    if (url.includes('mapbox-gl.css')) return 'css';
+
+    if (isMapboxHTTPFontsURL(url)) return 'font';
+    if (isMapboxHTTPSpriteURL(url)) return 'sprite';
+    if (isMapboxHTTPStyleURL(url)) return 'style';
+    if (isMapboxHTTPTileJSONURL(url)) return 'tilejson';
+
+    return 'other';
+}
+
+function getStyle(resourceTimers: Array<PerformanceEntry>): ?string {
+    if (resourceTimers) {
+        for (const timer of resourceTimers) {
+            const url = timer.name.split('?')[0];
+            if (isMapboxHTTPStyleURL(url)) {
+                const split = url.split('/').slice(-2);
+                if (split.length === 2) {
+                    return `${split[0]}:${split[1]}`;
```

Comment:
>Also, why get this from requests instead of from what is provided in the map constructor?

The map constructor data can be an inlined style object or the style may be set with `map.setStyle(...)`. Even if `setStyle` is a less likely code path, it's simpler to directly go through the resource requests.